### ISSUE TITLE
Chart update by calling `setState` in the point `mouseOver` event throws an exception

### ISSUE THREAD

**Issue by icecoldmax** [highcharts/highcharts-react#101]
Hi there,

I'm trying to use a mouseover event and show some of the hover event info outside the chart.

I've made an example here: https://stackblitz.com/edit/react-tzb1cr

You'll see that I am trying to use the mouseover to `setState()`, and then show the xAxis category underneath the chart. You'll see the console.log call works if you open up the console. I've commented out the setState as that is what breaks it. Try uncommenting it, then hovering over a point.

Am I doing it wrong? Thank you in advance!

**Comment by ppotaczek** [comment]
Hi @icecoldmax,

Thank you for reporting!

This issue indeed looks like a bug. Calling `setState()` in the `mouseOver` point event throws the error: `Cannot read property 'tooltipOptions' of null`. This problem does not seem to exist without the wrapper: http://jsfiddle.net/BlackLabel/vmw2d8gr/

To workaround set the `allowChartUpdate` option to `false`: https://stackblitz.com/edit/react-uoxbz7?file=index.js

Best regards!

**Comment by icecoldmax** [comment]
Thanks for getting back to me @ppotaczek !

I can confirm that `allowChartUpdate` option works. Unfortunately my use case is one where a single chart needs to change in response to different data being sent to it (imagine a chart with a bunch of filters and options in a sidebar that change the data to be displayed on the chart).

I've made a new stackblitz with an example: https://stackblitz.com/edit/react-yw38vy

You'll see that with `allowChartUpdate` set to true, the Update Series button will work, but the mouseover won't. When it's false, the mouseover works but I can't update the series.

I understand that this is a bug, but unfortunately I don't think I have the chops to solve it myself. Can you think of a workaround in the meantime, or perhaps an idea of when this bug might be addressed?

Thanks again!

**Comment by ppotaczek** [comment]
Hi @icecoldmax,

Sorry for the inconvenience. I am afraid that the fix may take some time, because the problem seems to be complicated and may require some changes in the Highcharts core.

However, you can make the `allowChartUpdate` flag dynamic, to solve both cases:

    constructor(props) {
      ...
      this.allowUpdate;
    }
  
    setHoverData = (e) => {
      this.allowUpdate = false;
      this.setState(...)
    }

    setNewSeriesData = () => {
      this.allowUpdate = true;
      this.setState(...);
    }
    render() {
      ...
      
      return (
        <div>
          <HighchartsReact
            ...
            allowChartUpdate={this.allowUpdate}
          />
        ...
        </div>
      )
    }

Live demo: https://stackblitz.com/edit/react-piyhi7?file=index.js

Best regards!

**Comment by icecoldmax** [comment]
Thanks @ppotaczek! That works great 👍 All the best with the fix.

As a side note, I notice there's no animation when the series changes (it just instantly updates). Is that related to the `allowChartUpdate` issue? I'm still familiarising myself with the library and the wrapper, but I notice that in [this issue](https://github.com/highcharts/highcharts-react/issues/15), the [final live demo](https://codesandbox.io/s/6xv0kr3v0n) has data being updated via a `setState()` as well, but the column growth is animated.

Thanks again for the quick and informative response!

**Comment by ppotaczek** [comment]
@icecoldmax,

Thanks, I am glad that I could help.

That animation problem is a Highcharts bug related to [the update with some event function](http://jsfiddle.net/BlackLabel/n3450xcg/). The issue is already fixed on the master branch: http://jsfiddle.net/BlackLabel/tbeswmaq/ and will be available in the next Highcharts version.

I think that a good practice is to keep all chart options in the state. When you call `setState` the options are overwritten and only the new ones are passed to the `chart.update` method. 
It turns out that this also solves your main problem because the options can be compared and the unnecessary update is not called.

    constructor(props) {
      ...
      this.state = {
        options: {...},
        ...
      };
    }
    setNewSeriesData = () => {
      ...
      this.setState({ 
        options: {...}
      });
    }
    render() {
     ...
      return (
        <div>
          <HighchartsReact
            ...
            options={this.state.options}
          />
        ...
        </div>
      )
    }

Live demo: https://stackblitz.com/edit/react-xemwjm?file=index.js

Best regards!

**Comment by icecoldmax** [comment]
@ppotaczek 

Wonderful! Setting the entire options hash in the state works perfectly.

Was that in the docs somewhere? My apologies if it was already documented and I just missed it 🙇

**Comment by ppotaczek** [comment]
I finally reproduced the issue without the wrapper: http://jsfiddle.net/BlackLabel/sfwz81o6/

It turns out that the problem was caused by the same bug that limited animation. So, as I mentioned before, it is already fixed on the master branch: http://jsfiddle.net/BlackLabel/86kbyoth/ and calling `setState` in point events will be possible in the next Highcharts release.

However, still the best way is to store all options in the state. The whole problem is avoided by updating the chart with only new options and that highly increases performance: https://stackblitz.com/edit/react-pvqjar?file=index.js


@icecoldmax, thanks for the valuable conversation! I will add to the docs a description of the optimal way to update.
